text/0000-multiarch-builders-and-package.md

```diff
@@ -0,0 +1,1119 @@
+
+- A new folder structure to organize the buildpacks binaries for multi-arch images, similar to this one:
+```bash
+# Option 1 - no variant is required
+.
+├── buildpack.toml                 // mandatory
+└── {os}                           // optional
+    └── {arch}                     // optional
+        └── bin
+            ├── build              // platform dependant binary (mandatory)
+            └── detect             // platform dependant binary (mandatory)
+
+# Option 2 - variant is required
+.
+├── buildpack.toml                  // mandatory
+└── {os}                            // optional
+    └── {arch}                      // optional
+        └── {variant}               // optional
+            ├── {name@version-1}    // optional
+            │ └── bin
+            │     ├── build         // platform dependant binary (mandatory)
+            │     └── detect        // platform dependant binary (mandatory)
+            └── {name@version-2}    // optional
+                └── bin
+                    ├── build       // platform dependant binary (mandatory)
+                    └── detect      // platform dependant binary (mandatory)
+```
+> **Note** 
+> For cross-compile buildpacks like Paketo, it looks easy to add a step to their Makefile to compile and separate the binaries following this structure. It is important to mention
+> that the final buildpack image will not change, this will only change the buildpack structure from `pack` perspective
+
+In case, this folder structure is not suitable for Buildpack Authors, **we propose** a new `path` attribute to be included 
+into the `target` section of the `buildpack.toml` to specify where the binaries are located in the filesystem.
+
+Based on the [RFC-0096](https://github.com/buildpacks/rfcs/blob/main/text/0096-remove-stacks-mixins.md) the new `buildpack.toml` schema will look like this:
+
+```toml
+[[targets]]
+os = "<operating system>"
+arch = "<system architecture>"
+variant = "<architecture variant>"
+
+# optional
+path = "<path to look for the binaries if the folder structure convention is not followed>" 
+
+[[targets.distributions]]
+name = "<distribution ID>"
+versions = ["<distribution version>"]
+```
+- When `more than 1 target is defined`
+  - When `--publish` or `--format file` is specified
+    - For each `target` an OCI image will be created, following these rules
+      - `pack` will copy everything in the root folder excluding all `{os}/**` folder 
+      - `pack` will determine the **target root binary folder**, this is the specific root folder for a given `target` (based on the `targets.path` in the buildpack.toml or inferring it from a folder structure similar to the one show above)
+        - `pack` will copy everything from the **target root binary folder** into the base buildpack package folder, in case a conflict file name is found, `pack` will **override** the existing file with the latest one
```

Comment:
Will these two copy steps outlined happen in different image layers? It seems like the idea here is to save disk space for files that are common across architectures. To achieve that, I think you'd need an image layer with the common files, that images for each architecture could include in their layer hierarchy. Maybe this was the idea and I missed it. If so, maybe worth mentioning somewhere.

Otherwise, if common files are in the same image layer as the architecture-specific files, clients would essentially be downloading and storing the common files once for each architecture.

Reply:
Ok, I did find the answer a bit further on: https://github.com/buildpacks/rfcs/pull/295/files#diff-cf9fc6f40b9a670966257d2475741958b49fb15684dd37cf8019514fe779885bR799

Reply:
If the common files are duplicated across architectures anyway, does the common file inheritance have much value? End users of buildpacks and builders won't really get much benefit. And the CNBs I work on (Heroku's), they won't have many (maybe 0) files shared across architectures.

Just thinking common file inheritance is a heavy feature with some potential gotchas, and buildpack authors who want/need to dedupe files could probably get pretty far just by symlinking or hardlinking files into multiple `{os}/{arch}` folders.

Reply:
Hi @joshwlewis 

Thank you very much for your feedback! I think that at least the `buildpack.toml` and the `package.toml` will be two common files shared across architectures, that's actually a win for a Buildpack Authors, right? because they do not need to keep separate files for each os/arch they want to support.

Reply:
> common file inheritance is a heavy feature with some potential gotchas

I agree. Maybe it makes sense to take it out of this RFC and add it later if buildpack authors want it

Reply:
> I think that at least the buildpack.toml and the package.toml will be two common files shared across architectures

I feel like `buildpack.toml` could be the one and only special case. IDK if `package.toml` is something that we even expect to be inside a buildpack root directory at build time. It's just easier to comprehend the packaging step if we don't move too much stuff around inside the directory tree. Perhaps we could spec out that we'd resolve symlinks when creating the layer if folks want to take advantage of that.

Reply:
Well, if I take a look at some buildpacks for Paketo for example, they have an structure like this one in the root folder:

```bash
├── LICENSE
├── NOTICE
├── README.md
├── buildpack.toml
├── package.toml
├── resources
│ ├── config.properties
```

In these cases, I think they want to shared those files on each platform images right?

Reply:
> In these cases, I think they want to shared those files on each platform images right?

They could copy, symlink, or hard link files into each platform folder. I don't think that's a big challenge for buildpack authors (they probably script this in some way anyway). Honestly, I think it might be a bit more straightforward for them to understand, too. Copy, symlink, and hardlink are established concepts; they won't need to read this RFC or the buildpack spec to know how to do that.

I still think there's convenience in this shared file logic. I'm mostly just thinking you could simplify the initial implementation, long term maintenance, and probably this RFC without it.